### PR TITLE
Improve mathtext.fontset docs and fix :mathmpl: cache bug.

### PR DESCRIPTION
**Pull request by anntzer**
In the mathtext tutorial, include all 5 fontsets now available (incl.
dejavusans/dejavuserif, which were previously missing) and generate the
expressions using mathmpl rather than static images.

This exposed three bugs:
- the MathTextParser cache previously did not get properly invalidated
  when rcParams["mathtext.fontset"] changed; so all parses (which depend
  on the font metrics) with different fontsets would be reused.
- mathmpl had not been updated to support dejavusans/dejavuserif.
- mathmpl would generate filenames not including the fontset info,
  causing different fontset renderings to overwrite each other.

Fix all of them.

See https://9831-7439715-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/tutorials/text/mathtext.html#fonts
vs the original
https://matplotlib.org/tutorials/text/mathtext.html#fonts

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/mathtext.py [hunk 1]
```diff
@@ -3320,27 +3320,28 @@ def __init__(self, output):
...
-        specifying the "default" font to use in the math expression,
-        used for all non-math text.
+        Parse the given math expression *s* at the given *dpi*.  If *prop* is
+        provided, it is a `.FontProperties` object specifying the "default"
+        font to use in the math expression, used for all non-math text.
 
         The results are cached, so multiple calls to :meth:`parse`
         with the same expression should be fast.
         """
+        return self._parse_cached(
```
Comment (timhoffm) [inline lib/matplotlib/mathtext.py]: Please add a comment that we have to defer to a seprate function and parse the rcParams in order to get poper caching behavior.

Comment (anntzer) [reply lib/matplotlib/mathtext.py]: done

File: lib/matplotlib/mathtext.py [hunk 2]
```diff
@@ -3320,27 +3320,31 @@ def __init__(self, output):
...
+        font to use in the math expression, used for all non-math text.
 
         The results are cached, so multiple calls to :meth:`parse`
         with the same expression should be fast.
         """
+        # lru_cache can't decorate parse() directly because the ps.useafm and
+        # mathtext.fontset rcParams also affect the parse (e.g. by affecting
+        # the glyph metrics).
+        return self._parse_cached(
+            s, dpi, prop, rcParams['ps.useafm'], rcParams['mathtext.fontset'])
```
Comment (QuLogic) [inline lib/matplotlib/mathtext.py]: Add `.lower()` here?

Comment (anntzer) [reply lib/matplotlib/mathtext.py]: Not needed, the rc validator (https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/rcsetup.py#L422) already enforces lowercase.

**Pull request merged**